ERCS/erc-xxxx.md

```diff
@@ -0,0 +1,243 @@
+Note that unlike `eth_call`, RIP-7560 transaction depends on its deployment and validation,
+so it can't be called in isolation from its deployment and validation
+If the transaction validation fails for any reason other than the failed signature check,
+returns an error object containing the details of the validation failure.
+
+Parameters:
+
+1. OBJECT - The RIP-7560 transaction object.
+   The `signature` field is optional.
+2. QUANTITY | TAG - integer block number, or the string "latest", "earliest", "pending", "safe" or "finalized"
+
+Returns:
+
+DATA - The return value of the `sender` execution frame.
+
+Error:
+
+DATA - The revert data of the first reverted frame.
+CODE - The error code indicating the type of error, which may include the entity that caused the revert on-chain.
+MESSAGE - The human-readable error that may include a decoding of the `DATA` field if possible.
+
+### Add RIP-7560 support for all remaining transaction-level RPC APIs
+
+This includes the following APIs: `eth_sendTransaction`, `eth_sendRawTransaction`,  `eth_getTransactionByHash`,
+`eth_getTransactionByBlockHashAndIndex`,
+`eth_getTransactionByBlockNumberAndIndex`.
+
+These methods have a very similar purpose and should support returning the new transaction type object.
+
+Note that the "transaction index position" is determined by the position of the transaction's **validation frame**.
+
+### Create a new JSON-RPC API method `eth_estimateRip7560TransactionGas`
+
+Performs a search for gas limit values required to make each of the frames of the RIP-7560 transaction execute
+successfully and without running out of gas.
+Note that for validation frames, only valid calls to an appropriate `AA_ENTRY_POINT` callback,
+such as `acceptAccount`, `acceptPaymaster`, `sigFailAccount` and `sigFailPaymaster`, is considered a success.
+
+If it fails to find such a value, returns an error message with the detailed description of the failure reason.
+
+Parameters:
+
+1. OBJECT - The RIP-7560 transaction object.
+   The `validationGasLimit`, paymasterValidationGasLimit, `paymasterGasLimit` and `callGasLimit` fields are optional.
+2. QUANTITY | TAG - integer block number, or the string "latest", "earliest", "pending", "safe" or "finalized"
+3. OBJECT - State override set
+   The State Override Set option allows you to change the state of a contract before executing the call. This means you
+   can modify the values of variables stored in the contract, such as balances and approvals for that call without
+   actually modifying the contract on the blockchain.
+   This behavior is equivalent to the one defined for the `eth_call` RPC method.
+
+Returns:
+
+1. Object
+
+| Name                        | Type     |
+|-----------------------------|----------|
+| validationGasLimit          | QUANTITY |
+| paymasterValidationGasLimit | QUANTITY |
+| paymasterPostOpGasLimit     | QUANTITY |
```

Comment:
postOpGasLimit is not something we should estimate: estimating would suggest that this estimation should be used - but that's wrong.
- We can calculate current value, but since it is not bound by opcode access, it can fail on different values.
- TokenPaymasters (and others) might "batch" calculations (e.g. swap), so estimation might be far less than actual gas - and revert.
- The entity that estimates is the account (sender), but the entity that pays is the paymaster, so the account has no incentive to do good estimation.
- Bottom line: I think paymasters should provide a constant value (via off chain service), and validate that in their paymasterValidation, and not rely on estimation at all.

Reply:
> Bottom line: I think paymasters should provide a constant value (via off chain service), and validate that in their paymasterValidation, and not rely on estimation at all.

What about paymasters that have no off-chain service?  E.g. a TokenPaymaster that caches lastKnownPrice in postOp, or occasionally calls Uniswap to rebalance its eth?  Where will the estimate come from in this case?

Reply:
1. Re: `since it is not bound by opcode access, it can fail on different values` - nothing is bound by opcode access. Opcode access is an opt-in protocol for bundlers, not part of the core protocol. Everything can fail all the time. This is exactly the situation with `estimateGas` in any other transaction type.
2. Re: `estimation might be far less than actual gas - and revert` - well that would not be a legitimate behaviour for the Paymaster. Suddenly requiring a large amount of gas out of the blue should be banned. A Paymaster that *might* require a huge post op gas can force it by adding `require(tx.postOpGasLimit > 1'000'000, "i need more gas")` in the `postOp`. Estimation algorithm should figure that out. This is the way to provide a constant gas limit value, if needed.
3. Re: `The entity that estimates is the account` - account is a smart contract and does not estimate anything. The Wallet Application calls the estimation. It is in the wallet's interest *not to* generate reverting transactions all the time.
4. Re: `not rely on estimation at all` - well, I don't think this is a viable solution for a smart contract of any kind.